File: SELF-Backend/SELF.Application/Applications/ApplicationService.cs
using Microsoft.EntityFrameworkCore;
using SELF.Application.Applications.DTOs;
using SELF.Application.Applications.Interfaces;
using SELF.Domain.Entities;
using SELF.Domain.Enums;
using SELF.Infrastructure.Data;
using SELF.Shared.Exceptions;

namespace SELF.Application.Applications;

public class ApplicationService(ApplicationDbContext context) : IApplicationService
{
    private readonly ApplicationDbContext _context = context;

    public async Task<ApplicationResponse> SubmitProposalAsync(Guid ngoId, ProjectProposalRequest request)
    {
        var ngo = await _context.NGOs.FirstOrDefaultAsync(n => n.Id == ngoId && !n.IsDeleted)
            ?? throw new NotFoundException("NGO organization not found.");

        var scheme = await _context.Schemes.FirstOrDefaultAsync(s => s.Id == request.SchemeId && !s.IsDeleted)
            ?? throw new NotFoundException("Selected scheme not found.");

        if (string.IsNullOrWhiteSpace(request.Title) || request.Budget.Count == 0)
        {
            throw new BadRequestException("Please provide a project title and at least one budget item.");
        }

        var randomNum = new Random().Next(100, 999);
        var appNumber = $"NGO-{DateTime.UtcNow.Year}-00{randomNum}";

        var totalBudgetAmount = request.Budget.Sum(b => b.Total > 0 ? b.Total : b.Quantity * b.UnitCost);
        var grantDisplay = $"₹{(totalBudgetAmount / 100000m):F2} Lakhs";

        var application = new Domain.Entities.Application
        {
            ApplicationNumber = appNumber,
            Title = request.Title.Trim(),
            SchemeId = scheme.Id,
            NgoId = ngo.Id,
            Status = ApplicationStatusType.ProposalSubmitted,
            GrantRequested = totalBudgetAmount,
            GrantRequestedDisplay = grantDisplay,
            SubmittedAt = DateTime.UtcNow
        };

        application.Project = new Project
        {
            Abstract = request.Abstract,
            State = request.State,
            District = request.District,
            Block = request.Block,
            Villages = request.Villages,
            MaleBeneficiaries = request.Beneficiaries.MaleCount,
            FemaleBeneficiaries = request.Beneficiaries.FemaleCount,
            TotalBeneficiaries = request.Beneficiaries.TotalCount,
            StBeneficiaries = request.Beneficiaries.StCount,
            Activities = request.Activities,
            ExpectedOutcomes = request.ExpectedOutcomes
        };

        application.Budget = new Budget
        {
            TotalAmount = totalBudgetAmount,
            Items = request.Budget.Select(b => new BudgetItem
            {
                Category = b.Category,
                Description = b.Description,
                Quantity = b.Quantity,
                UnitCost = b.UnitCost,
                Total = b.Total > 0 ? b.Total : b.Quantity * b.UnitCost
            }).ToList()
        };

        var todayStr = DateTime.UtcNow.ToString("yyyy-MM-dd");
        application.Reviews = new List<ApplicationReview>
        {
            new() { StageName = "Registered", Status = ReviewStatus.Completed, Remarks = "NGO registration verified.", ActionDate = DateTime.UtcNow.AddDays(-1), Sequence = 1 },
            new() { StageName = "Proposal Submitted", Status = ReviewStatus.Completed, Remarks = "Project Proposal submitted via online portal.", ActionDate = DateTime.UtcNow, Sequence = 2 },
            new() { StageName = "District Review", Status = ReviewStatus.Current, Remarks = "Awaiting inspection scheduler allocation.", Sequence = 3 },
            new() { StageName = "State Review", Status = ReviewStatus.Pending, Sequence = 4 },
            new() { StageName = "Approved", Status = ReviewStatus.Pending, Sequence = 5 },
            new() { StageName = "Sanctioned", Status = ReviewStatus.Pending, Sequence = 6 }
        };

        await _context.Applications.AddAsync(application);
        await _context.SaveChangesAsync();

        return new ApplicationResponse
        {
            Id = application.Id,
            ApplicationId = application.ApplicationNumber,
            Title = application.Title,
            SchemeName = scheme.Title,
            NgoName = ngo.Name,
            Status = "Proposal Submitted",
            SubmittedAt = todayStr,
            GrantRequested = grantDisplay,
            GrantRequestedAmount = totalBudgetAmount
        };
    }

    public async Task<List<ApplicationResponse>> GetNGOApplicationsAsync(Guid ngoId)
    {
        var apps = await _context.Applications
            .Include(a => a.Scheme)
            .Include(a => a.Ngo)
            .Where(a => a.NgoId == ngoId && !a.IsDeleted)
            .OrderByDescending(a => a.SubmittedAt)
            .ToListAsync();

        return apps.Select(a => new ApplicationResponse
        {
            Id = a.Id,
            ApplicationId = a.ApplicationNumber,
            Title = a.Title,
            SchemeName = a.Scheme?.Title ?? "",
            NgoName = a.Ngo?.Name ?? "",
            Status = a.Status.ToString(),
            SubmittedAt = a.SubmittedAt.ToString("yyyy-MM-dd"),
            GrantRequested = a.GrantRequestedDisplay,
            GrantRequestedAmount = a.GrantRequested
        }).ToList();
    }

    public async Task<List<ApplicationResponse>> GetAllApplicationsAsync()
    {
        var apps = await _context.Applications
            .Include(a => a.Scheme)
            .Include(a => a.Ngo)
            .Where(a => !a.IsDeleted)
            .OrderByDescending(a => a.SubmittedAt)
            .ToListAsync();

        return apps.Select(a => new ApplicationResponse
        {
            Id = a.Id,
            ApplicationId = a.ApplicationNumber,
            Title = a.Title,
            SchemeName = a.Scheme?.Title ?? "",
            NgoName = a.Ngo?.Name ?? "",
            Status = a.Status.ToString(),
            SubmittedAt = a.SubmittedAt.ToString("yyyy-MM-dd"),
            GrantRequested = a.GrantRequestedDisplay,
            GrantRequestedAmount = a.GrantRequested
        }).ToList();
    }

    public async Task<ApplicationTrackingResponse> TrackApplicationAsync(string applicationId)
    {
        var app = await _context.Applications
            .Include(a => a.Scheme)
            .Include(a => a.Ngo)
            .Include(a => a.Reviews)
            .FirstOrDefaultAsync(a => a.ApplicationNumber.ToUpper() == applicationId.Trim().ToUpper() && !a.IsDeleted)
            ?? throw new NotFoundException($"Application with ID \"{applicationId}\" was not found. Please verify the ID and try again.");

        var steps = app.Reviews.OrderBy(r => r.Sequence).Select(r => new TrackingStepDto
        {
            Name = r.StageName,
            Status = r.Status.ToString().ToLower(),
            Date = r.ActionDate?.ToString("yyyy-MM-dd"),
            Remarks = r.Remarks
        }).ToList();

        return new ApplicationTrackingResponse
        {
            ApplicationId = app.ApplicationNumber,
            NgoName = app.Ngo?.Name ?? "Registered NGO",
            SchemeName = app.Scheme?.Title ?? "Government Grant Scheme",
            Status = app.Status.ToString(),
            UpdatedAt = (app.UpdatedAt ?? app.SubmittedAt).ToString("yyyy-MM-dd"),
            Steps = steps
        };
    }

    public async Task<ApplicationResponse> UpdateApplicationStatusAsync(Guid id, ApplicationStatusType status, string? remarks)
    {
        var app = await _context.Applications
            .Include(a => a.Scheme)
            .Include(a => a.Ngo)
            .FirstOrDefaultAsync(a => a.Id == id && !a.IsDeleted)
            ?? throw new NotFoundException($"Application with ID '{id}' was not found.");

        app.Status = status;
        app.UpdatedAt = DateTime.UtcNow;

        if (!string.IsNullOrWhiteSpace(remarks))
        {
            var review = new ApplicationReview
            {
                ApplicationId = app.Id,
                StageName = status.ToString(),
                Status = ReviewStatus.Completed,
                ActionDate = DateTime.UtcNow,
                Remarks = remarks
            };
            _context.ApplicationReviews.Add(review);
        }

        await _context.SaveChangesAsync();

        return new ApplicationResponse
        {
            Id = app.Id,
            ApplicationId = app.ApplicationNumber,
            Title = app.Title,
            SchemeName = app.Scheme?.Title ?? "",
            NgoName = app.Ngo?.Name ?? "",
            Status = app.Status.ToString(),
            SubmittedAt = app.SubmittedAt.ToString("yyyy-MM-dd"),
            GrantRequested = app.GrantRequestedDisplay,
            GrantRequestedAmount = app.GrantRequested
        };
    }

    public async Task<bool> DeleteApplicationAsync(Guid id)
    {
        var app = await _context.Applications.FirstOrDefaultAsync(a => a.Id == id && !a.IsDeleted)
            ?? throw new NotFoundException($"Application with ID '{id}' was not found.");

        app.IsDeleted = true;
        app.UpdatedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();
        return true;
    }
}

File: self-main/src/services/applicationService.ts
import { api } from './api';
import type { ApiResponse } from './careerService';
import { APPLICATION_TRACKING_MOCKS, type ApplicationStatus } from '../constants/mockData';

export interface ProjectProposalPayload {
  title: string;
  schemeId: string;
  schemeName?: string;
  abstract: string;
  state: string;
  district: string;
  block?: string;
  villages?: string;
  beneficiaries: {
    maleCount: number;
    femaleCount: number;
    totalCount: number;
    stCount: number;
  };
  budget: {
    category: string;
    description: string;
    quantity: number;
    unitCost: number;
    total: number;
  }[];
  activities: string;
  expectedOutcomes: string;
  declarationChecked: boolean;
}

export interface ApplicationItem {
  id: string;
  applicationId: string;
  title: string;
  schemeName: string;
  ngoName: string;
  status: string;
  submittedAt: string;
  grantRequested: string;
  grantRequestedAmount: number;
}

export const applicationService = {
  // 1. Fetch all applications (for Admin & Officers)
  getAllApplications: async (): Promise<ApplicationItem[]> => {
    try {
      const response = await api.get<ApiResponse<ApplicationItem[]>>('/Application');
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
    } catch (e) {
      console.warn('Backend /Application unavailable, using fallback', e);
    }
    return [
      {
        id: 'f2cdfd4d-8391-4a01-bf18-0288261e577a',
        applicationId: 'NGO-2026-00124',
        title: 'Mobile Healthcare Clinic Jharkhand Phase 2',
        schemeName: 'Mobile Medical Units (MMU) in Scheduled Areas',
        ngoName: 'Gramin Vikas Parishad',
        status: 'Sanctioned',
        submittedAt: '2026-08-20',
        grantRequested: '₹40.00 Lakhs',
        grantRequestedAmount: 4000000,
      },
    ];
  },

  // 2. Update Application Status (Sanction, DistrictReview, Approved, Rejected)
  updateStatus: async (
    id: string,
    status: 'ProposalSubmitted' | 'DistrictReview' | 'StateReview' | 'Approved' | 'Sanctioned' | 'Rejected',
    remarks?: string
  ): Promise<ApplicationItem> => {
    const url = `/Application/${id}/status?status=${status}${remarks ? `&remarks=${encodeURIComponent(remarks)}` : ''}`;
    const response = await api.patch<ApiResponse<ApplicationItem>>(url);
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.message || 'Failed to update application status.');
    }
    return response.data.data;
  },

  // 3. Delete Application (Admin Only)
  deleteApplication: async (id: string): Promise<boolean> => {
    const response = await api.delete<ApiResponse<boolean>>(`/Application/${id}`);
    return Boolean(response.data.success);
  },

  // 4. Track an application by ID
  trackApplication: async (applicationId: string): Promise<ApplicationStatus> => {
    try {
      const response = await api.get<ApiResponse<any>>(`/Application/track/${applicationId}`);
      if (response.data.success && response.data.data) {
        const d = response.data.data;
        return {
          applicationId: d.applicationId,
          ngoName: d.ngoName,
          schemeName: d.schemeName,
          status: d.status,
          updatedAt: d.updatedAt,
          steps: d.steps || [],
        };
      }
    } catch (e) {
      console.warn('Backend tracking endpoint error, checking local store', e);
    }

    const matched = APPLICATION_TRACKING_MOCKS.find(
      (app) => app.applicationId.toUpperCase() === applicationId.trim().toUpperCase()
    );
    if (!matched) {
      throw new Error(`Application with ID "${applicationId}" was not found. Please verify the ID and try again.`);
    }
    return matched;
  },

  // 5. Submit a project proposal
  submitProposal: async (payload: ProjectProposalPayload): Promise<{ success: boolean; applicationId: string }> => {
    if (!payload.title || !payload.schemeId || payload.budget.length === 0) {
      throw new Error('Please fill in the project title, select a scheme, and build a project budget.');
    }

    try {
      const response = await api.post<ApiResponse<any>>('/Application/submit', payload);
      if (response.data.success && response.data.data) {
        return {
          success: true,
          applicationId: response.data.data.applicationId || response.data.data.applicationNumber,
        };
      }
    } catch (e: any) {
      console.warn('Backend submit failed or unavailable, using fallback', e);
      if (e.message && !e.message.includes('Network Error')) {
        throw e;
      }
    }

    // Fallback simulation
    const randomNum = Math.floor(100 + Math.random() * 900);
    const applicationId = `NGO-2026-00${randomNum}`;
    APPLICATION_TRACKING_MOCKS.push({
      applicationId,
      ngoName: 'Vikas Kalyan Sansthan',
      schemeName: payload.schemeName || 'State Grant Scheme',
      status: 'Proposal Submitted',
      updatedAt: new Date().toISOString().split('T')[0],
      steps: [
        { name: 'Registered', status: 'completed', date: new Date().toISOString().split('T')[0] },
        { name: 'Proposal Submitted', status: 'completed', date: new Date().toISOString().split('T')[0], remarks: 'Project Proposal submitted via online portal.' },
        { name: 'District Review', status: 'current', remarks: 'Awaiting inspection scheduler allocation.' },
        { name: 'State Review', status: 'pending' },
        { name: 'Approved', status: 'pending' },
        { name: 'Sanctioned', status: 'pending' },
      ],
    });

    return { success: true, applicationId };
  },

  // 6. Fetch all applications submitted by a specific NGO
  fetchNGOApplications: async (ngoId: string): Promise<ApplicationItem[]> => {
    try {
      const response = await api.get<ApiResponse<ApplicationItem[]>>(`/Application/ngo/${ngoId}`);
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
    } catch (e) {
      console.warn('Error fetching NGO applications', e);
    }
    return [
      {
        id: 'f2cdfd4d-8391-4a01-bf18-0288261e577a',
        applicationId: 'NGO-2026-00124',
        title: 'Mobile Healthcare Clinic Jharkhand Phase 2',
        schemeName: 'Mobile Medical Units (MMU) in Scheduled Areas',
        ngoName: 'Gramin Vikas Parishad',
        status: 'Sanctioned',
        submittedAt: '2026-08-20',
        grantRequested: '₹40.00 Lakhs',
        grantRequestedAmount: 4000000,
      },
    ];
  },
};
